python/ray/tests/test_reference_counting.py

```diff
@@ -236,15 +258,296 @@ def get_large_object(self):
+    ray.get(oid)
+
+    # Reference should be gone, check that array gets evicted.
+    _fill_object_store_and_get(array_oid_bytes, succeed=False)
+
+
+# Call a recursive chain of tasks that pass a serialized reference to the end
+# of the chain. The reference should still exist while the final task in the
+# chain is running and should be removed once it finishes.
+@pytest.mark.skip("Serialized ObjectID reference counting not implemented.")
+def test_recursive_serialized_reference(one_worker_100MiB):
+    @ray.remote
+    def recursive(ref, dep, max_depth, depth=0):
+        ray.get(ref[0])
+        if depth == max_depth:
+            return ray.get(dep[0])
+        else:
+            return recursive.remote(ref, dep, max_depth, depth + 1)
+
+    @ray.remote
+    def put():
+        return np.zeros(40 * 1024 * 1024, dtype=np.uint8)
+
+    max_depth = 5
+    array_oid = put.remote()
+    random_oid = ray.ObjectID.from_random()
+    head_oid = recursive.remote([array_oid], [random_oid], max_depth)
+
+    # Remove the local reference.
+    array_oid_bytes = array_oid.binary()
+    del array_oid
+
+    tail_oid = head_oid
+    for _ in range(max_depth - 1):
+        tail_oid = ray.get(tail_oid)
+
+    # Check that the remote reference pins the object.
+    _fill_object_store_and_get(array_oid_bytes)
+
+    # Fulfill the dependency, causing the tail task to finish.
+    ray.worker.global_worker.put_object(None, object_id=random_oid)
+    ray.get(tail_oid)
+
+    # Reference should be gone, check that array gets evicted.
+    _fill_object_store_and_get(array_oid_bytes, succeed=False)
+
+
+# Test that a passed reference held by an actor after the method finishes
+# is kept until the reference is removed from the actor. Also tests giving
+# the actor a duplicate reference to the same object ID.
+@pytest.mark.skip("Serialized ObjectID reference counting not implemented.")
+def test_actor_holding_serialized_reference(one_worker_100MiB):
+    @ray.remote
+    class GreedyActor(object):
+        def __init__(self):
+            pass
+
+        def set_ref1(self, ref):
+            self.ref1 = ref
+
+        def add_ref2(self, new_ref):
+            self.ref2 = new_ref
+
+        def delete_ref1(self):
+            self.ref1 = None
+
+        def delete_ref2(self):
+            self.ref2 = None
+
+    @ray.remote
+    def put():
+        return np.zeros(40 * 1024 * 1024, dtype=np.uint8)
+
+    # Test that the reference held by the actor isn't evicted.
+    array_oid = put.remote()
+    actor = GreedyActor.remote()
+    actor.set_ref1.remote([array_oid])
+
+    # Remove the local reference.
+    array_oid_bytes = array_oid.binary()
+    del array_oid
+
+    # Test that the remote reference still pins the object.
+    _fill_object_store_and_get(array_oid_bytes)
+
+    # Test that giving the same actor a duplicate reference works.
+    ray.get(actor.add_ref2.remote([ray.ObjectID(array_oid_bytes)]))
```

Comment:
Technically not possible to do this with the ray API, so maybe move this before we `del array_oid` and just pass in `[array_oid]` directly.